### DISCUSSION TITLE
TypeAdapter and slots in V2

### DISCUSSION THREAD

**Discussion by wolfmah** [pydantic/pydantic#5795] [Pydantic V2]
I am kinda new to the new features introduced in Python from the past few years (typing, dataclass and all).

I was fooling around with parsing JSON directly into dataclasses and started testing pydantic V2 to circumvent a limitation that was not present in V2 (seen in past issues in this repo).

I eventually stumble upon this: when using `TypeAdapter.validate_python` on a dataclass without slots, it succeeds. But with slots active, it fails.

```python
import dataclasses
import pydantic

@dataclasses.dataclass(slots=False)
class MyClass:
    a: str
    b: str

object_ = {"a": "hello", "b": "world"}
pydantic.TypeAdapter(MyClass).validate_python(object_)
print(MyClass(**object_))
```

The output of the succeeding `slots=False` is:
```
MyClass(a='hello', b='world')
```

When activating slots on the dataclass (`slots=True`), the output is
```
Traceback (most recent call last):
  File ".../dataclass.py", line 14, in <module>
    pydantic.TypeAdapter(MyClass).validate_python(object_)
  File ".../.venv/python-3.11.3/lib/python3.11/site-packages/pydantic/type_adapter.py", line 140, in validate_python
    return self.validator.validate_python(__object, strict=strict, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'MyClass' object has no attribute '__dict__'
```

I am not sure what to think of it, and I couldn't get far in the code of pydantic before getting lost.
Does anybody got any clues as to why this is? Am I trying something to do something that cannot be done and it just need a better error message? 🤔 

Python version
```
             pydantic version: 2.0a4
        pydantic-core version: 0.30.0 release build profile
                 install path: .../.venv/python-3.11.3/lib/python3.11/site-packages/pydantic
               python version: 3.11.3 (main, Apr 24 2023, 13:44:59) [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-13.3.1-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

**Comment by samuelcolvin** [comment]
Good catch, definitely looks like a bug.

What happens when you use the pydantic dataclasses decorator?

**Comment by samuelcolvin** [comment]
I've created in issue to track this #5797
